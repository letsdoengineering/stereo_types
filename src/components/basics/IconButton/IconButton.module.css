.button {
    border: none;
    background: none;
    cursor: pointer;
    outline: none;
    padding: 0;
}
.button-disabled {
    cursor: not-allowed;
}
.button-disabled .button-icon:hover {
    color: var(--dark-gray);
}

.button-icon:hover {
    color: var(--default-blue);
}

.button:focus {
    color: var(--default-blue);
    outline: 5px auto Highlight;
    outline: 5px auto -webkit-focus-ring-color;
}
