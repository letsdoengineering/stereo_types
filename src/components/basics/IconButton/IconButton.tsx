import React from 'react'
import getClassNames from '../../../utils/get-class-names'

import Icon, { IconProps } from '../Icon/Icon'
import * as iconButtonStyles from './IconButton.module.css'

export type IconButtonProps = IconProps &
  React.DetailedHTMLProps<React.ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement> & {
    /** label for accessibility reasons, not shown visually */
    label: string
    /** exposes className prop for the Icon component */
    iconClassName?: string
    /** exposes className prop for the button component */
    buttonClassName?: string
    /** function called when clicked */
    onClick(): void
    /** toggle for disabling the icon button*/
    disabled?: boolean
  }

/** IconButton: renders a button with a child icon and label that is not visually shown */
const IconButton: React.FC<IconButtonProps> = ({
  icon,
  iconSize = 'L',
  iconColor = 'Grey',
  iconClassName,
  buttonClassName,
  label,
  onClick,
  disabled = false,
  ...rest
}: IconButtonProps) => {
  const iconClassNames = getClassNames({
    defaultClasses: [iconButtonStyles.buttonIcon],
    className: iconClassName,
  })

  const buttonClassNames = getClassNames({
    defaultClasses: [iconButtonStyles.button],
    conditionalClasses: {
      [iconButtonStyles.buttonDisabled]: disabled,
    },
    className: buttonClassName,
  })

  const colour = disabled ? 'Grey' : iconColor
  return (
    <button
      disabled={disabled}
      type='button'
      onClick={onClick}
      className={buttonClassNames}
      {...rest}
    >
      <Icon icon={icon} iconSize={iconSize} iconColor={colour} className={iconClassNames} />
      <span className='visually-hidden'>{label}</span>
    </button>
  )
}

export default IconButton
