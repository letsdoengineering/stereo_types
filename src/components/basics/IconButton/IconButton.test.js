import * as React from 'react'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { axe } from 'jest-axe'

import IconButton from './IconButton'

describe('IconButton', () => {
  it('calls onClick when clicked', () => {
    const mockOnClick = jest.fn()
    render(<IconButton icon='times' label='an iconButton' onClick={mockOnClick} />)
    const icon = screen.getByText('an iconButton')
    userEvent.click(icon)
    expect(mockOnClick).toHaveBeenCalled()
  })

  it('has not changed the JSX layout of the page from previous snapshot', () => {
    const mockOnClick = jest.fn()
    const tree = renderer
      .create(<IconButton icon='CALENDAR' label='open calendar' onClick={mockOnClick} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('has not changed the JSX layout of the page from previous snapshot - disabled', () => {
    const mockOnClick = jest.fn()
    const tree = renderer
      .create(<IconButton disabled icon='CALENDAR' label='open calendar' onClick={mockOnClick} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders the page DOM that is accessible', async () => {
    const mockOnClick = jest.fn()
    const { container } = render(
      <IconButton icon='CALENDAR' label='open calendar' onClick={mockOnClick} />
    )
    const results = await axe(container)

    expect(results).toHaveNoViolations()
  })
})
