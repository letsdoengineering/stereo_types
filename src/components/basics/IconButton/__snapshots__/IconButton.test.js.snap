// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton has not changed the JSX layout of the page from previous snapshot - disabled 1`] = `
<button
  className="button buttonDisabled"
  disabled={true}
  onClick={[MockFunction]}
  type="button"
>
  <span
    aria-hidden="true"
    className="fas fa-CALENDAR iconSizeL iconColorGrey buttonIcon"
  />
  <span
    className="visually-hidden"
  >
    open calendar
  </span>
</button>
`;

exports[`IconButton has not changed the JSX layout of the page from previous snapshot 1`] = `
<button
  className="button"
  disabled={false}
  onClick={[MockFunction]}
  type="button"
>
  <span
    aria-hidden="true"
    className="fas fa-CALENDAR iconSizeL iconColorGrey buttonIcon"
  />
  <span
    className="visually-hidden"
  >
    open calendar
  </span>
</button>
`;
